Fullgraph graph capture with dynamo.

zhxchen17 opened the pull request:
Summary:
Following up on Avik's doc https://docs.google.com/document/d/11RW0Bbkp1QwFbEu8rCNW5d7wUFaEkxbL0uLyqcc2jTk/edit?tab=t.0

We are experimenting with a new API which utilizes torch.compile(fullgraph=True) and intend to use it to replace the old dynamo.export() API.

This PR adds a prototype for the API described in the doc.

Test Plan:
test_misc -- -k test_aot_capture

Rollback Plan:

Differential Revision: D79534608




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -2288,3 +2288,59 @@ def skip_code(code: types.CodeType) -> None:
+        artifacts = compiled_fn.get_artifacts()
+    """
+    from torch._dynamo.package import CompilePackage
+    package = CompilePackage(model)
```
Uhh, should you actually keep the original model live?

zhxchen17 commented:
To unblock @tugsbayasgalan from export experiments, I will just leave eval_frame there for now since imo our focus is on defining the core data structure for this PR. @ezyang mind reviewing this again?

zhxchen17 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -2288,3 +2288,79 @@ def skip_code(code: types.CodeType) -> None:
+        from torch._guards import TracingContext
+
+        backend_inputs[gm._backend_id] = BackendInput(
+            gm, example_inputs, TracingContext.get().fake_mode
```
@tugsbayasgalan I think fake mode can be useful for other consumers. Export as one consumer can easily get shape_env by fake_mode.shape_env

zhxchen17 committed:
Fullgraph graph capture with dynamo. (#159749)

Summary:

Following up on Avik's doc https://docs.google.com/document/d/11RW0Bbkp1QwFbEu8rCNW5d7wUFaEkxbL0uLyqcc2jTk/edit?tab=t.0

We are experimenting with a new API which utilizes torch.compile(fullgraph=True) and intend to use it to replace the old dynamo.export() API.

This PR adds a prototype for the API described in the doc.

Test Plan:
test_misc -- -k test_aot_capture

Rollback Plan:

Reviewed By: tugsbayasgalan

Differential Revision: D79534608

zhxchen17 commented:
@pytorchbot merge
